src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,545 @@
+    #[cfg_attr(miri, ignore)] // error: unsupported operation: can't call foreign function `TLS_method` on OS `linux`
+    async fn multi_part_upload_success() -> Result<(), S3MultiPartUploadError> {
+        let sdk_config = defaults().load().await;
+        let (bucket, path) = match s3_bucket_path_for_test() {
+            Some(tuple) => tuple,
+            None => return Ok(()),
+        };
+
+        let config = S3MultiPartUploaderConfig::default();
+        let mut uploader =
+            S3MultiPartUploader::try_new(&sdk_config, bucket.clone(), path.clone(), config).await?;
+
+        let expected_data = "onetwothree";
+        let _ = uploader.add_chunk(Bytes::from("one")).await?;
+        let _ = uploader.add_chunk(Bytes::from("two")).await?;
+        let _ = uploader.add_chunk(Bytes::from("three")).await?;
+
+        // This should trigger one single part upload.
+        let CompletedUpload {
+            part_count,
+            total_bytes_uploaded,
+        } = uploader.finish().await?;
+
+        // Getting the uploaded object from s3 and validating the contents.
+        let s3_client = s3::new_client(&sdk_config);
+        let uploaded_object = s3_client
+            .get_object()
+            .bucket(bucket)
+            .key(path)
+            .part_number(1) // fetching a particular part, so that the `parts_count` is populated in the result
+            .send()
+            .await
+            .unwrap();
+
+        let uploaded_parts_count: u32 = uploaded_object.parts_count().unwrap().try_into().unwrap();
+        assert!(uploaded_parts_count == part_count);
+
+        let body = uploaded_object.body.collect().await.unwrap().into_bytes();
+        assert!(body == expected_data);
+
+        let expected_bytes: u64 = Bytes::from(expected_data).len().try_into().unwrap();
+        assert!(total_bytes_uploaded == expected_bytes);
+
+        Ok(())
+    }
+
+    #[mz_ore::test(tokio::test(flavor = "multi_thread"))]
+    #[cfg_attr(coverage, ignore)] // https://github.com/MaterializeInc/materialize/issues/18898
+    #[cfg_attr(miri, ignore)] // error: unsupported operation: can't call foreign function `TLS_method` on OS `linux`
+    async fn multi_part_upload_buffer() -> Result<(), S3MultiPartUploadError> {
+        let sdk_config = defaults().load().await;
+        let (bucket, path) = match s3_bucket_path_for_test() {
+            Some(tuple) => tuple,
+            None => return Ok(()),
+        };
+
+        let config = S3MultiPartUploaderConfig::try_new(
+            ByteSize::mib(5).as_u64(),  // part_size_limit = 5MiB
+            ByteSize::mib(10).as_u64(), // file_size_limit = 10MiB
+        )?;
+        let mut uploader =
+            S3MultiPartUploader::try_new(&sdk_config, bucket.clone(), path.clone(), config).await?;
+
+        // Adding a chunk of 6MiB, should trigger an upload part since part_size_limit is 5MiB
+        let expected_data: &[u8] = &[97; 6291456]; // 6MiB
```

Comment:
I am generating 6MiB of data here, should I ignore this test? This test was useful to validate the buffer split.

Reply:
Yeah, seems fine. I'd use `vec![97; 6291456]` here though so you don't generate 6MB on the stack. Stacks are small.